exchanges/graphcache/src/operations/write.ts

```diff
@@ -343,19 +346,28 @@ const writeSelection = (
             ? InMemoryData.readLink(entityKey || typename, fieldKey)
             : undefined
         );
-        InMemoryData.writeLink(entityKey || typename, fieldKey, link);
+        const existingLink = InMemoryData.readLink(
+          entityKey || typename,
+          fieldKey
+        );
+        if (!isEqualLinkOrScalar(existingLink, link)) {
+          InMemoryData.writeLink(entityKey || typename, fieldKey, link);
+        }
       } else {
         writeField(ctx, getSelectionSet(node), ensureData(fieldValue));
       }
     } else if (entityKey && rootField === 'query') {
-      // This is a leaf node, so we're setting the field's value directly
-      InMemoryData.writeRecord(
+      const existingRecord = InMemoryData.readRecord(
         entityKey || typename,
-        fieldKey,
-        (fieldValue !== null || !getFieldError(ctx)
-          ? fieldValue
-          : undefined) as EntityField
+        fieldKey
       );
+      const value = (
+        fieldValue !== null || !getFieldError(ctx) ? fieldValue : undefined
+      ) as EntityField;
+      if (!isEqualLinkOrScalar(existingRecord, value)) {
```

Comment:
Are these checks too high in the code path?
We could in theory shift them down to `writeRecord` and `writeLink`, no? Or am I missing something?

Reply:
If we shift them down it also affects `squashLayer` which I wanted to avoid